### PR TITLE
VertexLoader: rewrite x64 JIT

### PR DESCRIPTION
**Pull request by Tilka**
~~The last commit should be dropped before this is merged.~~

### CODE REVIEW

**Comment by waddlesplash** [comment]
:clap: :clap: :clap:

Is this enabled by default?

**Comment by Tilka** [comment]
Yes, even if the architecture doesn't support it, hence some of the fails. :wink:

File: Source/Core/VideoCommon/VertexLoaderX64.cpp [hunk 1]
```diff
@@ -0,0 +1,458 @@
...
+	{
+		if (attribute == INDEX8)
+		{
+			MOVZX(64, 8, scratch1, data);
+		}
+		else
+		{
+			MOVZX(64, 16, scratch1, data);
+			// Convert to little-endian.
+			ROR(16, R(scratch1), Imm8(8));
```
Comment (magumagu) [inline Source/Core/VideoCommon/VertexLoaderX64.cpp]: We support BSWAP(16, R(X)) as an alias for rotate; might make the intent a bit more clear.  Also, minor performance tweak, you probably want to rotate then movzx to avoid the register merging penalty on Intel.

File: Source/Core/VideoCommon/VertexLoaderBase.cpp [hunk 2]
```diff
@@ -177,15 +231,24 @@ VertexLoaderBase* VertexLoaderBase::CreateVertexLoader(const TVtxDesc& vtx_desc,
 {
 	VertexLoaderBase* loader;
 
-#if 0
+#define COMPARE_VERTEXLOADERS
+
+#ifdef COMPARE_VERTEXLOADERS
```
Comment (waddlesplash) [inline Source/Core/VideoCommon/VertexLoaderBase.cpp]: Any point to this ifdef if it's unconditionally defined?

File: Source/Core/VideoCommon/VertexLoaderBase.cpp [hunk 3]
```diff
@@ -177,15 +231,24 @@ VertexLoaderBase* VertexLoaderBase::CreateVertexLoader(const TVtxDesc& vtx_desc,
 {
 	VertexLoaderBase* loader;
 
-#if 0
+#define COMPARE_VERTEXLOADERS
```
Comment (waddlesplash) [inline Source/Core/VideoCommon/VertexLoaderBase.cpp]: Reminder: undefine before merge!

File: Source/Core/VideoCommon/VertexLoaderX64.cpp [hunk 4]
```diff
@@ -0,0 +1,465 @@
...
+	int store_bytes = sizeof(float) * count_out;
+	if (load_bytes > 8)
+		MOVDQU(coords, data);
+	else if (load_bytes > 4)
+		MOVQ_xmm(coords, data);
+	else
+		MOVD_xmm(coords, data);
+	UpdateSourceOffset(attribute, load_bytes);
+
+	PSHUFB(coords, M(&shuffle_lut[format][count_in - 1]));
```
Comment (magumagu) [inline Source/Core/VideoCommon/VertexLoaderX64.cpp]: You need to guard the use of pshufb with an if ssse3.

Comment (Tilka) [inline Source/Core/VideoCommon/VertexLoaderX64.cpp]: That's in `IsInitialized()`.

File: Source/Core/VideoCommon/VertexLoaderX64.cpp [hunk 5]
```diff
@@ -0,0 +1,465 @@
...
+				MOV(32, R(scratch2), R(scratch1));
+			}
+			else
+			{
+				MOV(32, R(scratch3), R(scratch1));
+				SHL(32, R(scratch1), Imm8(16));
+				AND(32, R(scratch1), Imm32(0xF8000000));
+				MOV(32, R(scratch2), R(scratch1));
+
+				MOV(32, R(scratch1), R(scratch3));
```
Comment (magumagu) [inline Source/Core/VideoCommon/VertexLoaderX64.cpp]: (Repeat; I accidentally commented on an outdated commit.)  One of these MOVs is redundant.

File: Source/Core/VideoCommon/VertexLoaderX64.cpp [hunk 6]
```diff
@@ -0,0 +1,465 @@
...
+				SHUFPS(XMM0, R(XMM0), 0x45);
+				MOVUPS(MDisp(dst_reg, m_dst_ofs), XMM0);
+				m_dst_ofs += sizeof(float) * 4;
+			}
+		}
+	}
+
+	if (m_VtxDesc.PosMatIdx)
+	{
+		MOV(8, R(scratch1), MDisp(src_reg, posmatidx_ofs));
```
Comment (magumagu) [inline Source/Core/VideoCommon/VertexLoaderX64.cpp]: `MOVZX(32,8,...`; don't mix register sizes.

File: Source/Core/VideoCommon/VertexLoaderX64.cpp (hunk 4 above)
Comment (magumagu) [inline Source/Core/VideoCommon/VertexLoaderX64.cpp]: OK, sorry about the false alarm.

 BTW, IsInitialized is a terrible name...

Comment (Tilka) [inline Source/Core/VideoCommon/VertexLoaderX64.cpp]: Agreed. (But let's leave API changes for another PR.)

Comment (phire) [inline Source/Core/VideoCommon/VertexLoaderX64.cpp]: It wouldn't be too hard to make fallbacks.

The reason I say this, is because you have removed the old jit, so old Pre-Bulldozer AMDs (which many people annoyingly still use) will now fallback to the slower non-jit path.

**Comment by JMC47** [comment]
Windows is not working; seems to crash pretty much instantly in games.  On Linux, it seems to be working fine.

File: Source/Core/VideoCommon/VertexLoaderX64.cpp (hunk 4 above)
Comment (Tilka) [inline Source/Core/VideoCommon/VertexLoaderX64.cpp]: As mentioned on IRC: personally, I'm not interested in providing a fallback for `PSHUFB`.

File: Source/Core/VideoCommon/VertexLoaderX64.cpp (hunk 1 above)
Comment (Tilka) [inline Source/Core/VideoCommon/VertexLoaderX64.cpp]: But `data` is a memory location.

Comment (FioraAeterna) [inline Source/Core/VideoCommon/VertexLoaderX64.cpp]: In that case maybe load, then bswap, then movzx?

**Comment by JMC47** [comment]
Tales of Symphonia has a lot of green snow.  Fortune Street has green boards.

Super Smash Bros. Melee on Fountain of Dreams gives an unknown opcode error.

Super Smash Bros. Melee with 4 players on Dreamland seems slower than master for me?

**Comment by delroth** [comment]
@JMC47 care to have fifologs on fifoci for all these?

**Comment by shuffle2** [comment]
@JMC47 please try the current PR (with the debug commit), and replace RunVertices with this:
```c++
int RunVertices(int primitive, int count, DataReader src, DataReader dst) override
	{
		buffer_a.resize(count * a->m_native_vtx_decl.stride + 4);
		buffer_b.resize(count * b->m_native_vtx_decl.stride + 4);

		int count_a, count_b;
		{
			PROFILE("VertexLoader");
			count_a = a->RunVertices(primitive, count, src, DataReader(buffer_a.data(), buffer_a.data()+buffer_a.size()));
		}

		{
			PROFILE("VertexLoaderX64");
			count_b = b->RunVertices(primitive, count, src, DataReader(buffer_b.data(), buffer_b.data()+buffer_b.size()));
		}

		{
			PROFILE("profiling overhead");
		}

		if (count_a != count_b)
			ERROR_LOG(VIDEO, "Both vertexloaders have loaded a different amount of vertices (a: %d, b: %d).", count_a, count_b);

		if (memcmp(buffer_a.data(), buffer_b.data(), std::min(count_a, count_b) * m_native_vtx_decl.stride))
		{
			ERROR_LOG(VIDEO, "Both vertexloaders have loaded different data.");
			auto buf = std::make_unique<char[]>(buffer_a.size() * 4);
			auto pbuf = buf.get();
			pbuf += snprintf(pbuf, 3, "a:");
			for (size_t i = 0; i < buffer_a.size(); i++)
			{
				if (i % 4 == 0)
					pbuf += snprintf(pbuf, 2, " ");
				pbuf += snprintf(pbuf, 3, "%02x", buffer_a[i]);
			}
			ERROR_LOG(VIDEO, buf.get());

			buf = std::make_unique<char[]>(buffer_b.size() * 4);
			pbuf = buf.get();
			pbuf += snprintf(pbuf, 3, "b:");
			for (size_t i = 0; i < buffer_b.size(); i++)
			{
				if (i % 4 == 0)
					pbuf += snprintf(pbuf, 2, " ");
				pbuf += snprintf(pbuf, 3, "%02x", buffer_b[i]);
			}
			ERROR_LOG(VIDEO, buf.get());
		}

		memcpy(dst.GetPointer(), buffer_b.data(), count_b * m_native_vtx_decl.stride);
		m_numLoadedVertices += count;
		return count_b;
	}
```
make sure to put
```c++
#include "Common/StdMakeUnique.h"
```
in the file as well

File: Source/Core/VideoCommon/VertexLoaderX64.cpp [hunk 7]
```diff
@@ -0,0 +1,466 @@
...
+		 _mm_set_epi32(0xFFFFFFFFL, 0xFFFFFFFFL, 0xFFFF0203L, 0xFFFF0001L),  // 2x u16
+		 _mm_set_epi32(0xFFFFFFFFL, 0xFFFF0405L, 0xFFFF0203L, 0xFFFF0001L)}, // 3x u16
+		{_mm_set_epi32(0xFFFFFFFFL, 0xFFFFFFFFL, 0xFFFFFFFFL, 0x0001FFFFL),  // 1x s16
+		 _mm_set_epi32(0xFFFFFFFFL, 0xFFFFFFFFL, 0x0203FFFFL, 0x0001FFFFL),  // 2x s16
+		 _mm_set_epi32(0xFFFFFFFFL, 0x0405FFFFL, 0x0203FFFFL, 0x0001FFFFL)}, // 3x s16
+		{_mm_set_epi32(0xFFFFFFFFL, 0xFFFFFFFFL, 0xFFFFFFFFL, 0x00010203L),  // 1x float
+		 _mm_set_epi32(0xFFFFFFFFL, 0xFFFFFFFFL, 0x04050607L, 0x00010203L),  // 2x float
+		 _mm_set_epi32(0xFFFFFFFFL, 0x08090A0BL, 0x04050607L, 0x00010203L)}, // 3x float
+	};
+	static const __m128 scale_factors[32] = {
```
Comment (delroth) [inline Source/Core/VideoCommon/VertexLoaderX64.cpp]: Does that need alignment?

File: Source/Core/VideoCommon/VertexLoaderX64.cpp (hunk 4 above)
Comment (degasus) [inline Source/Core/VideoCommon/VertexLoaderX64.cpp]: I have to agree with flacs here. The removal of the jit in the old loader is only a ~2% slowdown. They also can't use the SSE intrinsic there, so it's likely even less. Through it's possible to write fallbacks, it's fine to optimize for the common case.

And sorry for the name "IsInitialized"

**Comment by JMC47** [comment]
@delroth They carry over through fifologs.  I record it on the broken build, it's broken in the fifolog on master, record on master, it works on the broken builds.

File: Source/Core/VideoCommon/VertexLoaderX64.cpp (hunk 7 above)
Comment (Tilka) [inline Source/Core/VideoCommon/VertexLoaderX64.cpp]: No, @shuffle2 quoted some documentation saying `__m128` is automatically 16-byte-aligned.

**Comment by Tilka** [comment]
Removed the `printf`s and the DEBUG commit.

**Comment by degasus** [comment]
There are still some bugs: https://fifoci.dolphin-emu.org/version/9938361b720ba77d35e0700f507cac0612e5346b/

**Comment by Tilka** [comment]
Fixed the green bug.

File: Source/Core/VideoCommon/VertexLoaderX64.cpp [hunk 8]
```diff
@@ -0,0 +1,453 @@
...
+	// Get the original count.
+	POP(32, R(ABI_RETURN));
+
+	if (m_VtxDesc.Position & MASK_INDEXED)
+	{
+		SUB(32, R(ABI_RETURN), R(skipped_reg));
+		RET();
+
+		SetJumpTarget(m_skip_vertex);
+		ADD(32, R(skipped_reg), Imm8(m_VtxDesc.Position == INDEX8 ? 1 : 2));
```
Comment (degasus) [inline Source/Core/VideoCommon/VertexLoaderX64.cpp]: This shall always increase by one. There is no need to skip two vertices if the index is 16 bit wide.

File: Source/Core/VideoCommon/VertexLoaderX64.cpp [hunk 9]
```diff
@@ -0,0 +1,453 @@
...
+
+		if (scaling_exponent)
+			MULPS(coords, M(&scale_factors[scaling_exponent]));
+	}
+
+	OpArg dest = MDisp(dst_reg, m_dst_ofs);
+	if (store_bytes > 8)
+		MOVUPS(dest, coords);
+	else
+		MOVLPS(dest, coords);
```
Comment (degasus) [inline Source/Core/VideoCommon/VertexLoaderX64.cpp]: What happens if we just want to store one value? else if (store_bytes > 4) ?

File: Source/Core/VideoCommon/VertexLoaderX64.cpp [hunk 10]
```diff
@@ -0,0 +1,453 @@
...
+				OR(32, R(scratch2), R(scratch1));
+
+				AND(32, R(scratch3), Imm8(0x0F));
+				OR(32, R(scratch2), R(scratch3));
+
+				MOV(32, R(scratch1), R(scratch2));
+				SHL(32, R(scratch1), Imm8(4));
+			}
+			OR(32, R(scratch1), R(scratch2));
+			SwapAndStore(32, MDisp(dst_reg, m_dst_ofs), scratch1);
```
Comment (degasus) [inline Source/Core/VideoCommon/VertexLoaderX64.cpp]: isn't this swapping ond both loading and saving + the swizzling a bit redundant?

File: Source/Core/VideoCommon/VertexLoaderX64.cpp [hunk 11]
```diff
@@ -0,0 +1,453 @@
...
+			break;
+	}
+	if (attribute == DIRECT)
+		m_src_ofs += load_bytes;
+}
+
+void VertexLoaderX64::GenerateVertexLoader()
+{
+	m_native_components = 0;
+	memset(&m_native_vtx_decl, 0, sizeof(m_native_vtx_decl));
```
Comment (degasus) [inline Source/Core/VideoCommon/VertexLoaderX64.cpp]: This should be moved into the base class.

File: Source/Core/VideoCommon/VertexLoaderX64.cpp [hunk 12]
```diff
@@ -0,0 +1,453 @@
...
+		if (tm[i])
+			texmatidx_ofs[i] = m_src_ofs++;
+	}
+
+	OpArg data = GetVertexAddr(ARRAY_POSITION, m_VtxDesc.Position);
+	ReadVertex(data, m_VtxDesc.Position, m_VtxAttr.PosFormat, m_VtxAttr.PosElements + 2, 3, m_VtxAttr.PosFrac, &m_native_vtx_decl.position);
+
+	if (m_VtxDesc.Normal)
+	{
+		static const u8 map[] = {7, 6, 15, 14, 30};
```
Comment (degasus) [inline Source/Core/VideoCommon/VertexLoaderX64.cpp]: Where does this numbers come from?

Comment (degasus) [inline Source/Core/VideoCommon/VertexLoaderX64.cpp]: float aren't scaled, so the last number doesn't matter

File: Source/Core/VideoCommon/VertexLoaderX64.cpp [hunk 13]
```diff
@@ -0,0 +1,453 @@
...
+					m_dst_ofs += sizeof(float);
+				MOVSS(MDisp(dst_reg, m_dst_ofs), XMM0);
+				m_dst_ofs += sizeof(float);
+			}
+			else
+			{
+				m_native_vtx_decl.texcoords[i].components = 4;
+				m_native_vtx_decl.texcoords[i].offset = m_dst_ofs;
+				m_native_vtx_decl.texcoords[i].type = VAR_FLOAT;
+				m_native_vtx_decl.texcoords[i].integer = false;
```
Comment (degasus) [inline Source/Core/VideoCommon/VertexLoaderX64.cpp]: I'd move the m_native_vtx_decl.texcoords[i].enable also here. For tc[i] == true, it's alredy set.

**Comment by degasus** [comment]
I can't review x64 assembly, but all video common changes LGTM

File: Source/Core/VideoCommon/VertexLoaderX64.cpp [hunk 14]
```diff
@@ -0,0 +1,453 @@
...
+			u8 scaling_exponent = m_VtxAttr.texCoord[i].Frac;
+			ReadVertex(data, tc[i], m_VtxAttr.texCoord[i].Format, elements, elements, scaling_exponent, &m_native_vtx_decl.texcoords[i]);
+			m_native_components |= VB_HAS_UV0 << i;
+		}
+		if (tm[i])
+		{
+			m_native_components |= (VB_HAS_TEXMTXIDX0 | VB_HAS_UV0) << i;
+			m_native_vtx_decl.texcoords[i].enable = true;
+			MOVZX(64, 8, scratch1, MDisp(src_reg, texmatidx_ofs[i]));
+			PXOR(XMM0, R(XMM0));
```
Comment (FioraAeterna) [inline Source/Core/VideoCommon/VertexLoaderX64.cpp]: Is this pxor really needed? I'm not sure it is

File: Source/Core/VideoCommon/VertexLoaderX64.cpp [hunk 15]
```diff
@@ -0,0 +1,453 @@
...
+	// TODO: load constants into registers outside the main loop
+
+	const u8* loop_start = GetCodePtr();
+
+	u32 posmatidx_ofs = 0;
+	if (m_VtxDesc.PosMatIdx)
+		posmatidx_ofs = m_src_ofs++;
+
+	u32 texmatidx_ofs[8];
+	const u64 tm[8] = {
```
Comment (FioraAeterna) [inline Source/Core/VideoCommon/VertexLoaderX64.cpp]: Could this be a BitSet8? Do we need tm[i] and friends for any purpose other than boolean if statements? It might be cleaner/simpler to do it that way.

File: Source/Core/VideoCommon/VertexLoaderX64.cpp [hunk 16]
```diff
@@ -0,0 +1,453 @@
...
+	static const u32 mask_0f  = 0x0F0F0F0F;
+	static const u32 mask_f0  = 0xF0F0F0F0;
+	static const u32 mask_fc  = 0xFCFCFCFC;
+	switch (format)
+	{
+		case FORMAT_24B_888:
+		case FORMAT_32B_888x:
+		case FORMAT_32B_8888:
+			MOV(32, R(scratch1), data);
+			if (format != FORMAT_32B_8888 || !elements)
```
Comment (FioraAeterna) [inline Source/Core/VideoCommon/VertexLoaderX64.cpp]: I remember this line had a comment in the C implementation; maybe duplicate/reference it here? Since it was a bit odd

File: Source/Core/VideoCommon/VertexLoaderX64.cpp [hunk 17]
```diff
@@ -0,0 +1,453 @@
...
+			break;
+
+		case FORMAT_16B_4444:
+			//                   RRRRGGGG BBBBAAAA
+			// AAAAAAAA BBBBBBBB GGGGGGGG RRRRRRRR
+			LoadAndSwap(16, scratch1, data);
+			if (cpu_info.bBMI2)
+			{
+				PDEP(32, scratch2, scratch1, M(&mask_0f));
+				PDEP(32, scratch1, scratch1, M(&mask_f0));
```
Comment (FioraAeterna) [inline Source/Core/VideoCommon/VertexLoaderX64.cpp]: I just want to say I love you for this because this is the first time I've ever seen someone find a use for pdep and that is a sad shame

File: Source/Core/VideoCommon/VertexLoaderX64.cpp (hunk 9 above)
Comment (FioraAeterna) [inline Source/Core/VideoCommon/VertexLoaderX64.cpp]: I don't think that happens, from my memory; iirc every use of that function involves store_bytes == 2 or 3.

File: Source/Core/VideoCommon/VertexLoaderX64.cpp [hunk 18]
```diff
@@ -0,0 +1,453 @@
...
+		// Sign extend
+		if (format == FORMAT_BYTE)
+			PSRAD(coords, 24);
+		if (format == FORMAT_SHORT)
+			PSRAD(coords, 16);
+
+		CVTDQ2PS(coords, R(coords));
+
+		if (scaling_exponent)
+			MULPS(coords, M(&scale_factors[scaling_exponent]));
```
Comment (FioraAeterna) [inline Source/Core/VideoCommon/VertexLoaderX64.cpp]: It would be super cool if we could keep this in a register; iirc it doesn't change? but I might be wrong

File: Source/Core/VideoCommon/VertexLoaderX64.cpp [hunk 19]
```diff
@@ -0,0 +1,453 @@
...
+			BSWAP(16, scratch1);
+			MOVZX(64, 16, scratch1, R(scratch1));
+		}
+		if (array == ARRAY_POSITION)
+		{
+			CMP(attribute == INDEX8 ? 8 : 16, R(scratch1), Imm8(-1));
+			m_skip_vertex = J_CC(CC_E, true);
+		}
+		MOVZX(64, 32, scratch2, M(&g_main_cp_state.array_strides[array]));
+		IMUL(64, scratch1, R(scratch2));
```
Comment (FioraAeterna) [inline Source/Core/VideoCommon/VertexLoaderX64.cpp]: Do we really need a 64-bit multiply here? I kinda doubt the array strides can be that large given the size of the GC memory space...

If the answer is yes, we really do need a 64-bit multiply, could you change array_strides to a u64 or ptrdiff_t to turn this into an IMUL with memory, saving an op? Otherwise you can just merge the two without doing that, if it's a 32-bit multiply.

File: Source/Core/VideoCommon/VertexLoaderX64.cpp [hunk 20]
```diff
@@ -0,0 +1,453 @@
...
+		m_native_vtx_decl.posmtx.enable = true;
+		m_native_vtx_decl.posmtx.offset = m_dst_ofs;
+		m_native_vtx_decl.posmtx.type = VAR_UNSIGNED_BYTE;
+		m_native_vtx_decl.posmtx.integer = true;
+		m_dst_ofs += sizeof(u32);
+	}
+
+	// Prepare for the next vertex.
+	ADD(64, R(dst_reg), Imm32(m_dst_ofs));
+	const u8* cont = GetCodePtr();
```
Comment (FioraAeterna) [inline Source/Core/VideoCommon/VertexLoaderX64.cpp]: I think this can use a FixupBranch

File: Source/Core/VideoCommon/VertexLoaderX64.cpp [hunk 21]
```diff
@@ -0,0 +1,453 @@
...
+		{
+			m_native_components |= (VB_HAS_TEXMTXIDX0 | VB_HAS_UV0) << i;
+			m_native_vtx_decl.texcoords[i].enable = true;
+			MOVZX(64, 8, scratch1, MDisp(src_reg, texmatidx_ofs[i]));
+			PXOR(XMM0, R(XMM0));
+			CVTSI2SS(XMM0, R(scratch1));
+			if (tc[i])
+			{
+				m_native_vtx_decl.texcoords[i].components = 3;
+				if (elements == 1)
```
Comment (FioraAeterna) [inline Source/Core/VideoCommon/VertexLoaderX64.cpp]: are you sure this is right? iirc with elements == 1 we write one float, and with elements == 2 we write two floats, one of which is zero (kind of like in elements==4 we write 4 floats, three of which are zero).

File: Source/Core/VideoCommon/VertexLoaderX64.cpp (hunk 19 above)
Comment (degasus) [inline Source/Core/VideoCommon/VertexLoaderX64.cpp]: The stride is the most 0xff, so an u8 would be enough

File: Source/Core/VideoCommon/VertexLoaderX64.cpp (hunk 15 above)
Comment (degasus) [inline Source/Core/VideoCommon/VertexLoaderX64.cpp]: I'm sorry, the second idea isn't possible as this is a bitset. But we could already convert it within the vertexloaderbase.

File: Source/Core/VideoCommon/VertexLoaderX64.cpp (hunk 10 above)
Comment (Tilka) [inline Source/Core/VideoCommon/VertexLoaderX64.cpp]: Yes, but it keeps the code readable and let's me use an 8-bit immediate :3

File: Source/Core/VideoCommon/VertexLoaderX64.cpp (hunk 8 above)
Comment (Tilka) [inline Source/Core/VideoCommon/VertexLoaderX64.cpp]: Thanks for noticing, this doesn't make sense at all.

File: Source/Core/VideoCommon/VertexLoaderX64.cpp (hunk 9 above)
Comment (Tilka) [inline Source/Core/VideoCommon/VertexLoaderX64.cpp]: You are both right. I added the use of this function for storing just one float value.

File: Source/Core/VideoCommon/VertexLoaderX64.cpp (hunk 19 above)
Comment (FioraAeterna) [inline Source/Core/VideoCommon/VertexLoaderX64.cpp]: Multiplies smaller than 32-bit aren't any faster (often even slower...) so it's pobably best to stick with a 32-bit multiply.

Comment (Tilka) [inline Source/Core/VideoCommon/VertexLoaderX64.cpp]: We also wouldn't get the automatic zero-extension which is needed for the `MRegSum` operand. So yeah, 32 bit.

File: Source/Core/VideoCommon/VertexLoaderX64.cpp (hunk 18 above)
Comment (Tilka) [inline Source/Core/VideoCommon/VertexLoaderX64.cpp]: It doesn't change but there are different values of `scaling_exponent`. I'll leave this optimization for later.

File: Source/Core/VideoCommon/VertexLoaderX64.cpp (hunk 17 above)
Comment (Tilka) [inline Source/Core/VideoCommon/VertexLoaderX64.cpp]: Thanks a lot :D

File: Source/Core/VideoCommon/VertexLoaderX64.cpp (hunk 21 above)
Comment (Tilka) [inline Source/Core/VideoCommon/VertexLoaderX64.cpp]: The code was correct but I refactored it anyway.

**Comment by Tilka** [comment]
Rebased onto @degasus' PR for vertex skipping in the software render. This should result in no FifoCI diffs anymore. All known bugs fixed, all comments addressed. Ready to merge from my side.

**Comment by delroth** [comment]
@dolphin-emu-bot rebuild

**Comment by delroth** [comment]
@dolphin-emu-bot rebuild

**Comment by delroth** [comment]
@dolphin-emu-bot rebuild

File: Source/Core/VideoCommon/VertexLoader.cpp [hunk 22]
```diff
@@ -195,16 +130,22 @@ void VertexLoader::CompileVertexTranslator()
...
+		m_native_vtx_decl.posmtx.components = 4;
+		m_native_vtx_decl.posmtx.enable = true;
+		m_native_vtx_decl.posmtx.offset = nat_offset;
+		m_native_vtx_decl.posmtx.type = VAR_UNSIGNED_BYTE;
+		m_native_vtx_decl.posmtx.integer = true;
+		nat_offset += 4;
 		m_VertexSize += 1;
 	}
 
+
```
Comment (degasus) [inline Source/Core/VideoCommon/VertexLoader.cpp]: empty line

**Comment by delroth** [comment]
@dolphin-emu-bot rebuild

**Comment by degasus** [comment]
LGTM, @FioraAeterna Do you still know some issues?

**Comment by Nerboruto** [comment]
this pr break mario kart double dash
image

**Comment by erisdev** [comment]
this change broke compatibility with the intel iris 5100 graphics chip for some games (notably paper mario) that were working fine before. is there any chance of doing hardware capabilities detection and using the old implementation if the harware-specific features the new implementation requires are not available? i'd like to not be stuck with an old version.

**Comment by delroth** [comment]
@erisdev please file a bug on bugs.dolphin-emu.org. This change should have 0 driver-specific impact.

**Comment by degasus** [comment]
@erisdev please do not file a bug report. This change has not included the old bounding box implementation, so if you lack D3D11 or OpenGL4.3 hardware features, BBox won't work any more. Honestly, your GPU should support both. Have you updated your driver? Did you try both the OGL and D3D video backend?

**Comment by erisdev** [comment]
@degasus i'm on a mac so i don't get the luxury of updating the drivers or trying direct3d, i get what apple gives me.
